addons/escoria-core/game/core-scripts/esc_room.gd

```diff
@@ -89,31 +112,110 @@ func _ready():
 	
 	if global_id.empty():
 		global_id = name
+	
+	# Determine whether this room was run from change_scene or directly
+	if escoria.main.has_node(name):
+		is_run_directly = false
+	else:
+		is_run_directly = true
+	
+	perform_script_events()
+
+
+func perform_script_events():
+	if esc_script and escoria.event_manager._running_event == null \
+	or (escoria.event_manager._running_event != null \
+	and escoria.event_manager._running_event.name != "load"):
 		
-	if esc_script:
-		run_script_event("setup")
-		var rc = yield(escoria.event_manager, "event_finished")
-		while rc[1] != "setup":
-			rc = yield(escoria.event_manager, "event_finished")
-		if rc[0] != ESCExecution.RC_OK:
-			return rc[0]
-		
+		# Manage player location at room start
 		if (escoria.globals_manager.get_global("ESC_LAST_SCENE") == null \
 			or escoria.globals_manager.get_global("ESC_LAST_SCENE").empty()) \
 			and player != null \
 			and escoria.object_manager.get_start_location() != null:
 			player.teleport(escoria.object_manager.get_start_location().node)
+		
+		# If the room was loaded from change_scene and automatic transitions
+		# are not disabled, do the transition out now
+		if enabled_automatic_transitions \
+		and not is_run_directly \
+		and not exited_previous_room:
+			var script_transition_out = escoria.esc_compiler.compile([
+				":transition_out",
+				"transition %s out" % ProjectSettings.get_setting(
+					"escoria/ui/default_transition"
+				),
+				"hide_menu main"
+			])
+			escoria.event_manager.queue_event(
+				script_transition_out.events['transition_out']
+			)
 			
 		
-		escoria.main.scene_transition.transition()
-		yield(escoria.main.scene_transition, "transition_done")
-	
-		run_script_event("ready")
-		rc = yield(escoria.event_manager, "event_finished")
+		# Run the setup event
+		_run_script_event("setup")
+		
+		if enabled_automatic_transitions \
+		or (
+			not enabled_automatic_transitions \
+			and escoria.globals_manager.get_global("BYPASS_LAST_SCENE")
+		):
+			var script_transition_in = escoria.esc_compiler.compile([
+				":transition_in",
+				"transition %s in" % ProjectSettings.get_setting(
+					"escoria/ui/default_transition"
+				),
+				"wait 0.1"
+			])
+			escoria.event_manager.queue_event(script_transition_in.events['transition_in'])
+			
+		# Run the ready event
+		if escoria.event_manager._running_event == null \
+		or (escoria.event_manager._running_event != null \
+		and escoria.event_manager._running_event.name != "load"):
+			_run_script_event("ready")
+		
+		# Wait for ready event to be done
+		var rc = yield(escoria.event_manager, "event_finished")
 		while rc[1] != "ready":
```

Comment:
Hmmmm... this might never be true, because the room has no ready event, right?

Reply:
Indeed. I added a check to verify if the script has this event in its script, and don't yield if it doesn't.